core/unix/signal.c

```diff
@@ -2408,7 +2395,10 @@ handle_sigprocmask(dcontext_t *dcontext, int how, kernel_sigset_t *app_set,
         check_signals_pending(dcontext, info);
     }
     if (!execute_syscall) {
-        handle_post_sigprocmask(dcontext, how, app_set, oset, sigsetsize);
+        int status = handle_post_sigprocmask(dcontext, how, app_set, oset, sigsetsize);
```

Comment:
But we still have the problems pointed out in the prior review, don't we?  If the syscall fails b/c of a bad old-sigset, the whole thing should abort, right?  So we need to undo the changes to info->app_sigblocked, and we might need to check old-sigset before calling check_signals_pending if that's a non-undoable change -- or we have to convince ourselves by looking at the ordering of things that we don't actually deliver a signal there and it's a functional nop (but wastes time in dispatch looking for a signal to deliver when they're all blocked).

Reply:
> But we still have the problems pointed out in the prior review, don't we? If the syscall fails b/c of a bad old-sigset, the whole thing should abort, right?

As I mentioned in the previous comment thread, we don't need to abort. The syscall has this weird non-atomic behavior where the new sigset still takes hold, even if the syscall fails with EFAULT due to bad old sigset.  

> So we need to undo the changes to info->app_sigblocked,

No need to undo. The changes are made successfully natively too. 

As DR's record of the signals blocked in the app is correct, I don't think we need any changes in how we handle signals.

Reply:
OK, my only request then would be to document this kernel behavior in the DR code and not just the test.

Reply:
Yes, documented already.